### PR TITLE
metadata: set the index on meta_data as UNIQUE

### PR DESCRIPTION
**Pull request by phweyland**
fixes a root cause of #7780

EDIT: If duplicates before migration keeps the oldest ones.

### CODE REVIEW

**Comment by HansBull** [comment]
Humm...who'se 32 will get merged first (or at all)...?

**Comment by phweyland** [comment]
> Humm...who'se 32 will get merged first (or at all)...?

I think this one is important, but it can wait (I'll resolve the conflicts if necessary) :smiley:

**Comment by HansBull** [comment]
Important, yes. Sans doute.

File: src/common/database.c [hunk 1]
```diff
@@ -1712,6 +1712,24 @@ static int _upgrade_library_schema_step(dt_database_t *db, int version)
     sqlite3_exec(db->handle, "COMMIT", NULL, NULL, NULL);
     new_version = 31;
   }
+  else if(version == 31)
+  {
+    sqlite3_exec(db->handle, "BEGIN TRANSACTION", NULL, NULL, NULL);
+
+    // remove duplicates
+    TRY_EXEC("DELETE FROM main.meta_data WHERE rowid NOT IN (SELECT min(rowid) "
```
Comment (TurboGit) [inline src/common/database.c:1720]: I see that now and indeed you took exactly the fix I proposed, that is smallest rowid. You've then double checked that this gets the displayed values, right ?

Comment (phweyland) [reply src/common/database.c:1720]: > I see that now and indeed you took exactly the fix I proposed, that is smallest rowid. You've then double checked that this gets the displayed values, right ?

I've checked on sqlstudio with faulty data.
I've also checked with db without duplicate (and got it untouched :) ) but not doubled checked with faulty data. Give me five minutes.

Comment (phweyland) [reply src/common/database.c:1720]: Double checked.

**Comment by phweyland** [comment]
> I would include `value` too in the key.

In the index ? What would be the benefit ?

**Comment by TurboGit** [comment]
Ok, thanks! Duplicated records are properly removed on my side.

**Comment by TurboGit** [comment]
Merged in master and queued for 3.4.1.

**Comment by HansBull** [comment]
> > I would include `value` too in the key.
> 
> In the index ? What would be the benefit ?

2 steps less for the query in metadata.c line 457 and one less for the query line 465, accordig to `EXPLAIN`

**Comment by phweyland** [comment]
> > > I would include `value` too in the key.
> > 
> > 
> > In the index ? What would be the benefit ?
> 
> 2 steps less for the query in metadata.c line 457 and one less for the query line 465, accordig to `EXPLAIN`

I don't know (yet !) this sqlite keyword. I have to dig into it ...
This said, these two queries return one or few records. Fetching the value once the record is found may be a additional step, but it costs almost nothing. 
Maintaining the index has also a cost. So there is a balance somewhere.
But these EXPLAIN information may be key for queries which manipulate potentially a huge number of records like in collect / collection and where every step is repeated thousands of times. I'll give a go ! Thanks

**Pull request merged**